### PR TITLE
Renable tests

### PR DESCRIPTION
**Pull request by bramtayl**
The only tests that are currently running in CI are google tests and the tests in the commandline folder, and they are only running on Linux. This leaves out the tests in the regression folder, and the tests in the soak folder. It also leaves out the examples in the manual. This enables them on all platforms. I also added checksums for all .csd files and split up the integration tests into separate ctests. See the documentation for how checksums are useful. Spliting up the integration tests is useful because:

- You can manually change the status of a test
- You can use ctest to only run 1 test, or only run previously failing tests, etc.
- ctest keeps track of the amount of time it takes to run each test

In terms of performance, tests take less than a minute to run, which seems manageable to me. You can run the tests with the commands:

```
cd build
ctest
```

If you would like to exclude longer-running tests, run `ctest -LE soak`, which only takes 10 seconds or so to run

There are two commits: the first is just all the new checksums, and the second is the substantive changes

### CODE REVIEW

**Comment by bramtayl** [comment]
Good idea to merge this ASAP I think because without it most things are untested [even with it, test coverage is only like 40%]

**Comment by stekyne** [comment]
Why is there 3000 changes?

**Comment by bramtayl** [comment]
Two things: 1) moved all the csound files in the test folder into one place 2) added checksums for all of them

**Comment by kunstmusik** [comment]
I do not like this PR. All of the tests were organized by their purpose and now they are all mixed in one folder. Also, python_tests would sound like testing the python csound API rather than what this is, just using python as a test runner. And what is the purpose of the checksums? This PR has zero description about what is intended by the PR and discussion with the developer group would have made sense prior to a large reorganization like this.

**Comment by vlazzarini** [comment]
agreed, tests were organised in a sound way for development. Not sure why we needed any changes. It's confusing to me.

**Comment by bramtayl** [comment]
>  All of the tests were organized

Oh? How were they organized before? I saw three folders, `soak`, `regression`, and `commandline`, with `commandline` having some csound files in it and four subfolders, `array`, `contrib`, `structs`, and `udo`? Maybe I'm failing to see the logic?

Happy to organize the tests back into folders too!

> python_tests would sound like testing the python csound API

Oh yeah, I can change the name!

> what is the purpose of the checksums

Oh, sorry, I should have explained more. The original python code calculated checksums for the generated wav files too. The idea is to generate a wav file with a csound file, and save the checksum of that wav file. Then, if we make a change that changes the wav file, then the checksum will change and the tests will fail.

> Not sure why we needed any changes

Well, the tests weren't running on CI at all before. And if you don't test your code with CI, people can break it very easily.

**Comment by bramtayl** [comment]
Ok, renamed python tests to command-line tests, and separated into two commits. The first commit is the substantive changes, and the second commit is reorganizing etc.

**Comment by bramtayl** [comment]
Ok, kept all the csound files in their original folders too. The second commit is just checksums now

**Comment by bramtayl** [comment]
Does that address all the concerns?

File: CMakeLists.txt [hunk 1]
```diff
@@ -110,6 +112,38 @@ option(REQUIRE_PTHREADS "For non-Windows systems, set whether Csound will use th
...
+if (BUILD_TESTS)
+    # c tests require the static library
+    set(BUILD_STATIC_LIBRARY ON)
+endif()
+
+if (TRACK_COVERAGE)
+    check_cxx_compiler_flag(-g has_debug_info)
+    if (has_debug_optimization)
+        # disable optimizations for better coverage stats
+        add_compile_options(-Og)
```
Comment (stekyne) [inline CMakeLists.txt]: I'm not sure I like this. Someone enabling coverage would not expect all of a sudden that the binary output would be extremely slow. I think the coverage build should be a separate target that is built separately in the CI process rather potentially affecting the main binary output.

File: CMakeLists.txt [hunk 2]
```diff
@@ -110,6 +112,38 @@ option(REQUIRE_PTHREADS "For non-Windows systems, set whether Csound will use th
...
+    endif()
+
+    # link option is required for the compiler option, so set and unset
+    set(CMAKE_REQUIRED_LINK_OPTIONS --coverage)
```
Comment (stekyne) [inline CMakeLists.txt]: Is this a Gcc/Clang flag? It won't work on Windows in that case

File: InOut/CMakeLists.txt [hunk 3]
```diff
@@ -86,13 +86,9 @@ if(APPLE)
...
 endif()
 
-if(USE_ALSA AND LINUX)
-    if (USE_VCPKG)
-        find_package(ALSA MODULE REQUIRED)
-    else()
-        find_package(ALSA REQUIRED)
-    endif()
-
+find_package(ALSA)
```
Comment (stekyne) [inline InOut/CMakeLists.txt]: This will use the MODULE search instead of CONFIG then no?

File: generate_coverage.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# add an empty coverage folder
+# set TRACK_COVERAGE=ON
```
Comment (stekyne) [inline generate_coverage.sh]: This is a CMAKE var but this comment doesn't really look like it. It makes it sound like an ENV var no?

File: Python/CMakeLists.txt [hunk 5]
```diff
@@ -2,8 +2,6 @@
 
 option(INSTALL_PYTHON_INTERFACE "Build the Python interface" ON)
 
-find_package(Python3)
```
Comment (stekyne) [inline Python/CMakeLists.txt]: Why is this removed?

File: tests/c/CMakeLists.txt [hunk 6]
```diff
@@ -20,34 +20,22 @@ if(BUILD_TESTS)
...
-            set(CMAKE_CXX_FLAGS "-stdlib=libc++")
-        endif()
-
-        target_include_directories(unittests PRIVATE
-            ${GTEST_INCLUDE_DIRS})
+        ${CMAKE_SOURCE_DIR}/H
+        ${CMAKE_SOURCE_DIR}/interfaces
+    )
+    
+    set_property(TARGET unittests PROPERTY CXX_STANDARD 14)
```
Comment (stekyne) [inline tests/c/CMakeLists.txt:27]: Should be `target_link_options` maybe?

**Comment by stekyne** [comment]
Added some more comments. Sorry for the delays but this is a huge PR.

The checksum idea is interesting but it will be a bit of a maintenance burden. Unless the wave file is bit for bit identical, it will break the test, the developer will have to generate the checksums again? There is similar concept in the reactjs world of snapshot tests. They really have little value and devs will end up just regenerating the checksum blindly. You mentioned the checksum feature existed before your changes but where were they stored? I'm not really familiar with this part of the code so that would be good to know.

I'm still not sure merging the unit tests and the integration tests (CSDs) is a good idea. The unit tests mostly test the engine and C code. The CSD/python tests are more like integration or e2e testing. I think if a developer is editing code and wants check if tests are passing, they don't want to run the entire suite of 2000 tests. 

Also the coverage stuff seems somewhat unrelated. I think that could have been another PR entirely and that also warrants some discussion. In general, I think it only works on linux/osx machines. I've commented about the optimisation issue. That's a side effect best avoided. Code coverage should probably be a CI thing that only needs to run once and not per platform. Maybe we can add a new GH action workflow just for that.

File: Python/CMakeLists.txt (hunk 5 above)
Comment (bramtayl) [reply Python/CMakeLists.txt]: I just moved it to the start of the main CMake file so we can use python from anywhere in the code?

File: InOut/CMakeLists.txt (hunk 3 above)
Comment (bramtayl) [reply InOut/CMakeLists.txt]: Yes, but there is no ALSA module so nothing to worry about I don't think

File: CMakeLists.txt (hunk 2 above)
Comment (bramtayl) [reply CMakeLists.txt]: Yes. I only added it temporarily so the line below would work on gcc/clang though. I remove it two lines later?

File: generate_coverage.sh (hunk 4 above)
Comment (bramtayl) [reply generate_coverage.sh]: Ok, I'll make a note in the comment!

File: CMakeLists.txt (hunk 1 above)
Comment (bramtayl) [reply CMakeLists.txt]: > coverage build should be a separate target that is built separately in the CI process

Hmm, what do you mean?

File: tests/c/CMakeLists.txt (hunk 6 above)
Comment (bramtayl) [reply tests/c/CMakeLists.txt:27]: Hmm, what do you mean?

**Comment by bramtayl** [comment]
> Sorry for the delays but this is a huge PR

Np, thanks for the feedback!

> Unless the wave file is bit for bit identical, it will break the test, the developer will have to generate the checksums again?

Yes. I did add an python option to regenerate the checksums here: 

https://github.com/csound/csound/blob/a57cc69dd83804d57d189f69998d9ab1eed051ea/tests/test.py.in#L8

>  where were they stored?

Soak had some checksums: https://github.com/csound/csound/blob/a1580f9cdf331c35dceb486f4231871ce0b00266/tests/soak/Sample_CheckSums#L1

>  they don't want to run the entire suite of 2000 tests

I can see that, but all the tests only take 5 min to run. They could of course set this to off: https://github.com/csound/csound/blob/a57cc69dd83804d57d189f69998d9ab1eed051ea/CMakeLists.txt#L115. We could also leave it off by default. And that time is pretty easy to cut down too by upping the tempo in the longer-running csound files

> Also the coverage stuff seems somewhat unrelated

Agreed, can put that into a separate PR if you want

>  I think it only works on linux/osx machines

Yup, a bit unfortunate

> Maybe we can add a new GH action workflow just for that

Great idea!

File: InOut/CMakeLists.txt (hunk 3 above)
Comment (stekyne) [reply InOut/CMakeLists.txt]: https://cmake.org/cmake/help/latest/module/FindALSA.html There is the built in one for CMake

File: CMakeLists.txt (hunk 1 above)
Comment (stekyne) [reply CMakeLists.txt]: Build a separate CMake target just for the coverage so there is no chance of contaminating the main build. It probably should exclude the opcodes actually and just test the engine. Also you can specify what compile options like no optimisation. It should probably be just a debug build I guess.

File: Python/CMakeLists.txt (hunk 5 above)
Comment (stekyne) [reply Python/CMakeLists.txt]: But is it needed by anything else bar the Python stuff for Csound?

File: tests/c/CMakeLists.txt (hunk 6 above)
Comment (stekyne) [reply tests/c/CMakeLists.txt:27]: https://cliutils.gitlab.io/modern-cmake/chapters/features/cpp11.html should set the standard via the `target_compile_features` method I mean. Not the link_options sorry

Comment (bramtayl) [reply tests/c/CMakeLists.txt:27]: Ok, changed it to that!

File: Python/CMakeLists.txt (hunk 5 above)
Comment (bramtayl) [reply Python/CMakeLists.txt]: Yeah, we need python for the command-line tests too

File: generate_coverage.sh (hunk 4 above)
Comment (bramtayl) [reply generate_coverage.sh]: Removed coverage PR to follow

File: InOut/CMakeLists.txt (hunk 3 above)
Comment (bramtayl) [reply InOut/CMakeLists.txt]: Oh yeah. Well, I'm not sure how this change snuck in because it isn't relevant anyway, so got rid ofit.

File: CMakeLists.txt (hunk 1 above)
Comment (bramtayl) [reply CMakeLists.txt]: Removed coverage, separate PR to follow

File: tests/c/CMakeLists.txt (hunk 6 above)
Comment (bramtayl) [reply tests/c/CMakeLists.txt:27]: Nvm, tried it and it didn't work, so reverted

**Comment by bramtayl** [comment]
Removed coverage stuff for now

File: CMakeLists.txt (hunk 1 above)
Comment (bramtayl) [reply CMakeLists.txt]: But comments: I think a separate target would be overkill, but I can add a clear warning in the option description? I definitely don't think it would make sense to exclude opcodes: the whole point is to see which opcodes are tested and which aren't.

File: Python/CMakeLists.txt (hunk 5 above)
Comment (stekyne) [reply Python/CMakeLists.txt]: Then it should only be found if building tests etc. We should keep these find_package calls as local as possible. You can have multiple if required.

File: tests/c/CMakeLists.txt (hunk 6 above)
Comment (stekyne) [reply tests/c/CMakeLists.txt:27]: https://github.com/csound/csound/pull/1786/files#diff-7424ebd04f7f2f0b6a6f91f4a25c6fec73f2f5cc33d1b519db3cb4523dfd16c3R38 I've used this elsewhere and it works. In fact, maybe let me merge the changes I've proposed there and see what is applicable after

File: tests/commandline/nested_strings.checksum
Comment (stekyne) [inline tests/commandline/nested_strings.checksum:1]: What is this checksum for? It doesn't seem to be an audio sample?

File: cmake/cmake-utilities.cmake [hunk 7]
```diff
@@ -126,4 +126,13 @@ macro(assign_bool variable)
      else()
          set(${variable} OFF)
      endif()
-endmacro()
\ No newline at end of file
+endmacro()
+
+# add a csound test based on a .csd file
+function(add_commandline_test filename status)
```
Comment (stekyne) [inline cmake/cmake-utilities.cmake]: Are these statuses predefined somewhere? Why are some listed as `passing`, `no_output` or `indeterminate_output`?

**Comment by stekyne** [comment]
> > Unless the wave file is bit for bit identical, it will break the test, the developer will have to generate the checksums again?
> 
> Yes. I did add an python option to regenerate the checksums here:
> 
> https://github.com/csound/csound/blob/a57cc69dd83804d57d189f69998d9ab1eed051ea/tests/test.py.in#L8
> 
> > where were they stored?
> 
> Soak had some checksums:
> 
> https://github.com/csound/csound/blob/a1580f9cdf331c35dceb486f4231871ce0b00266/tests/soak/Sample_CheckSums#L1
> 

Ok but why split the checksums out to a separate file? I think just one file with a list of them is sufficient. It looks like they haven't changed in 10 years. Also the switch to having all tests manually defined. Can you explain the rational for doing that?

> > they don't want to run the entire suite of 2000 tests
> 
> I can see that, but all the tests only take 5 min to run. They could of course set this to off:
> 
> https://github.com/csound/csound/blob/a57cc69dd83804d57d189f69998d9ab1eed051ea/CMakeLists.txt#L115
> 
> . We could also leave it off by default. And that time is pretty easy to cut down too by upping the tempo in the longer-running csound files

5 mins is far too long for unit tests tbh. It should be on the order of seconds to get quick feedback that any code changes have broken something or not. Integration tests that involve multiple process starts/stops etc are always going to be slow. They should be separated out from the unit test runner.

**Comment by bramtayl** [comment]
> the switch to having all tests manually defined

There are a couple of advantages here:

- You can manually change the status of a test (for example to "non-deterministic output", or "different output on windows", etc.)
- You can use ctest to only run 1 test, or only run previously failing tests, etc.
- ctest keeps track of the amount of time it takes to run each test

> why split the checksums out to a separate file

Well, see above about why it makes sense to have a separate ctest for each test. For each separate ctest, we could parse one file with all the checksums in it over and over. But that would be a waste of resources I think. And the csound files are already scattered around several different folders anyways

> 5 mins is far too long for unit tests

Well, I suppose we could leave `ADD_COMMANDLINE_TESTS` off by default and only enable it for CI. But again, even with the commandline tests, the current tests only cover 40% of the code, which means 60% is untested. And that is frankly terrifying.

File: cmake/cmake-utilities.cmake (hunk 7 above)
Comment (bramtayl) [reply cmake/cmake-utilities.cmake]: See here: https://github.com/csound/csound/blob/89205ff3ba5d3d28e92933feebe37263169c2443/tests/test.py.in#L32

File: tests/commandline/nested_strings.checksum
Comment (bramtayl) [reply tests/commandline/nested_strings.checksum:1]: See here: https://github.com/csound/csound/blob/89205ff3ba5d3d28e92933feebe37263169c2443/tests/commandline/nested_strings.csd#L1

File: Python/CMakeLists.txt (hunk 5 above)
Comment (bramtayl) [reply Python/CMakeLists.txt]: Ok will do

Comment (bramtayl) [reply Python/CMakeLists.txt]: Done

File: tests/test.py.in [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+# built plugins should be in the same folder as the csound binary
+os.environ['OPCODE6DIR64'] = "$<TARGET_FILE_DIR:csound-bin>"
+os.environ['CS_USER_PLUGINDIR'] = "$<TARGET_FILE_DIR:csound-bin>"
+
+# directory: path where csd file is
+# filename: e.g. test, if file is test.csd
+# status: one of "passing", "indeterminate_output", or "no_output"
+#
+# - "no_output": generates no output
+# - "stalls": runs indefinitely or for too long
```
Comment (stekyne) [inline tests/test.py.in]: Any test that stalls or never completes is just a broken test. I don't think we should accept that as a valid test. Did you notice any tests doing that?

File: tests/commandline/nested_strings.checksum
Comment (stekyne) [reply tests/commandline/nested_strings.checksum:1]: But what is the checksum representing? I'm not really clear on that. The actual csd file itself?

File: cmake/cmake-utilities.cmake (hunk 7 above)
Comment (stekyne) [inline cmake/cmake-utilities.cmake]: I think the name `add_commandline_test ` is not very clear what is the purpose. The command line word here is immaterial as most things are done via the CLI. It is really adding an integration test. I think `add_integration_test` is better.

File: tests/test.py.in (hunk 8 above)
Comment (stekyne) [reply tests/test.py.in]: You should comment them out in the cmake and leave a note to prevent it running rather than having a bail out option here imo.

**Comment by stekyne** [comment]
> > the switch to having all tests manually defined
> 
> There are a couple of advantages here:
> 
> * You can manually change the status of a test (for example to "non-deterministic output", or "different output on windows", etc.)
> * You can use ctest to only run 1 test, or only run previously failing tests, etc.
> * ctest keeps track of the amount of time it takes to run each test
> 

Ok that sounds pretty reasonable. I'll play around with running tests a bit more and see how it feels.

> > why split the checksums out to a separate file
> 
> Well, see above about why it makes sense to have a separate ctest for each test. For each separate ctest, we could parse one file with all the checksums in it over and over. But that would be a waste of resources I think. And the csound files are already scattered around several different folders anyways
> 

Ok but I would move them to a checksum folder I think. Should avoid having massive folders of files. Are you testing if there is a checksum missing? It fails the test I presume? I would also document how this all needs to line up. There is a requirement the filename of the csd and checksum line up right?

> > 5 mins is far too long for unit tests
> 
> Well, I suppose we could leave `ADD_COMMANDLINE_TESTS` off by default and only enable it for CI. But again, even with the commandline tests, the current tests only cover 40% of the code, which means 60% is untested. And that is frankly terrifying.

Not sure `ADD_COMMANDLINE_TESTS` is the best name here. `BUILD_TESTS` is the existing var that was used, how come you are not using it? I think that should be the main one. We can add separate ones for `BUILD_UNIT_TESTS` vs `BUILD_INTEGRATION_SUITE` maybe. To disable certain parts of the test suite.

**Comment by kunstmusik** [comment]
I still don't find this PR useful. It's taken all of the commandline tests and moved them to CMake, but runs each with a separate invocation of python. It now makes it not possible to run the c unit tests separately from the main tests, nor separate from the longer running tests that we didn't want to run everytime we test. It's made test running so slow we couldn't use them during development. Previously if we needed to run multiple sets of tests, we'd just all all of the targets (i.e., `make test csdtests etc.`). I see this PR as a step backwards. 

Additionally, I asked for more information and the description of this PR still has essentially no information that would aid any developer reviewing the PR as to the purpose of the set of changes.

**Comment by vlazzarini** [comment]
and I still don't understand what these changes are supposed to fix.

File: cmake/cmake-utilities.cmake (hunk 7 above)
Comment (bramtayl) [reply cmake/cmake-utilities.cmake]: Ok, changed it to that

File: tests/test.py.in (hunk 8 above)
Comment (bramtayl) [reply tests/test.py.in]: Ok, done!

**Comment by bramtayl** [comment]
> It now makes it not possible to run the c unit tests separately from the main tests, nor separate from the longer running tests

Ok, I added labels based on the three folders, so that you can run `ctest -L regression` etc.

**Comment by bramtayl** [comment]
> I asked for more information

I added documentation here: https://github.com/csound/csound/blob/84647b483ae9d28ec5ce757bcd3918423794e284/cmake/cmake-utilities.cmake#L131 . Does that help?

> I still don't understand what these changes are supposed to fix

Here's a blog post I found online about why you should test code with CI: https://www.copado.com/resources/blog/the-ci-cd-pipeline-why-testing-is-required-at-every-stage

**Comment by kunstmusik** [comment]
> > I asked for more information
> 
> I added documentation here:
> 
> https://github.com/csound/csound/blob/84647b483ae9d28ec5ce757bcd3918423794e284/cmake/cmake-utilities.cmake#L131
> 
> . Does that help?
> > I still don't understand what these changes are supposed to fix
> 
> Here's a blog post I found online about why you should test code with CI: https://www.copado.com/resources/blog/the-ci-cd-pipeline-why-testing-is-required-at-every-stage

The question isn't about running tests with CI, we already had that for years. The question is what is the point of this PR? Your PR description is still this:

image
 
which explains nothing for anyone trying to review the PR. The PR title says it is trying to re-enable tests but the contents of the PR is more of a refactor with little regard for understanding of the existing test system. 

* What does moving the tests to CMake get us?
* What are the performance differences in running the tests?
* What should someone who has been doing development with the existing system expect when using these changes? What will they need to do differently for running and adding tests than before? 

The current test system could certainly be improved, but so far this PR has been a lot of steps backwards.

**Comment by bramtayl** [comment]
> The question isn't about running tests with CI, we already had that for years

I think if you examine the CI output you'll find that the only tests that are currently running in CI are the tests in the `commandline` folder, and they are only running on Linux, not windows or mac. This leaves out the google tests, the tests in the `regression` folder, and the tests in the `soak` folder.

**Comment by bramtayl** [comment]
> PR description is still this

Ok, updated the description, hope that helps!

**Comment by bramtayl** [comment]
Added the example csd files from the manual too. This brings test coverage from like 10% to like 50%, tests take 3-5 min to run on CI

**Comment by bramtayl** [comment]
Changed to check midi output checksums too. So five commits:

1) Add checksums (for midi output and wav output)
2) Add examples: add the contents of the examples folder from the manual
3) Delete tests with opcodes that don't seem to exist any longer
4) Remove file flags: mostly remove `-odac` because we don't need real-time output
5) The core changes

**Comment by bramtayl** [comment]
Ok, got a little lost here, but simplified back down again. Tests take less than a minute to run

**Comment by bramtayl** [comment]
Part of #1846
